runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/GroupByKeyTranslatorBatch.java

```diff
@@ -17,74 +17,264 @@
-    implements TransformTranslator<
-        PTransform<PCollection<KV<K, V>>, PCollection<KV<K, Iterable<V>>>>> {
+    extends GroupingTranslator<K, V, Iterable<V>, GroupByKey<K, V>> {
+
+  /** Literal of binary encoded Pane info. */
+  private static final Expression PANE_NO_FIRING = lit(toByteArray(NO_FIRING, PaneInfoCoder.of()));
+
+  /** Defaults for value in single global window. */
+  private static final List<Expression> GLOBAL_WINDOW_DETAILS =
+      windowDetails(lit(new byte[][] {EMPTY_BYTE_ARRAY}));
+
+  private boolean useCollectList = true;
+
+  public GroupByKeyTranslatorBatch() {}
+
+  public GroupByKeyTranslatorBatch(boolean useCollectList) {
+    this.useCollectList = useCollectList;
+  }
 
   @Override
-  public void translateTransform(
-      PTransform<PCollection<KV<K, V>>, PCollection<KV<K, Iterable<V>>>> transform,
-      AbstractTranslationContext context) {
-
-    @SuppressWarnings("unchecked")
-    final PCollection<KV<K, V>> inputPCollection = (PCollection<KV<K, V>>) context.getInput();
-    Dataset<WindowedValue<KV<K, V>>> input = context.getDataset(inputPCollection);
-    WindowingStrategy<?, ?> windowingStrategy = inputPCollection.getWindowingStrategy();
-    KvCoder<K, V> kvCoder = (KvCoder<K, V>) inputPCollection.getCoder();
-    Coder<V> valueCoder = kvCoder.getValueCoder();
-
-    // group by key only
-    Coder<K> keyCoder = kvCoder.getKeyCoder();
-    KeyValueGroupedDataset<K, WindowedValue<KV<K, V>>> groupByKeyOnly =
-        input.groupByKey(KVHelpers.extractKey(), EncoderHelpers.fromBeamCoder(keyCoder));
-
-    // group also by windows
-    WindowedValue.FullWindowedValueCoder<KV<K, Iterable<V>>> outputCoder =
-        WindowedValue.FullWindowedValueCoder.of(
-            KvCoder.of(keyCoder, IterableCoder.of(valueCoder)),
-            windowingStrategy.getWindowFn().windowCoder());
-    Dataset<WindowedValue<KV<K, Iterable<V>>>> output =
-        groupByKeyOnly.flatMapGroups(
-            new GroupAlsoByWindowViaOutputBufferFn<>(
-                windowingStrategy,
-                new InMemoryStateInternalsFactory<>(),
-                SystemReduceFn.buffering(valueCoder),
-                context.getSerializableOptions()),
-            EncoderHelpers.fromBeamCoder(outputCoder));
-
-    context.putDataset(context.getOutput(), output);
+  public void translate(GroupByKey<K, V> transform, Context cxt) {
+    WindowingStrategy<?, ?> windowing = cxt.getInput().getWindowingStrategy();
+    TimestampCombiner tsCombiner = windowing.getTimestampCombiner();
+
+    Dataset<WindowedValue<KV<K, V>>> input = cxt.getDataset(cxt.getInput());
+
+    KvCoder<K, V> inputCoder = (KvCoder<K, V>) cxt.getInput().getCoder();
+    KvCoder<K, Iterable<V>> outputCoder = (KvCoder<K, Iterable<V>>) cxt.getOutput().getCoder();
+
+    Encoder<V> valueEnc = cxt.valueEncoderOf(inputCoder);
+    Encoder<K> keyEnc = cxt.keyEncoderOf(inputCoder);
+
+    // In batch we can ignore triggering and allowed lateness parameters
```

Comment:
in batch we can ignore allowed lateness but not triggers as triggers can be set in processing time in global window (See nexmark query 12 that sets a trigger repeatedly after processingTime pass the first element in
pane + a given period)

Reply:
We've already briefly discussed that ... processing time in batch is something totally different than processing time in streaming. In my eyes this trigger makes no sense at all in the batch case: processing time is meaningless in batch as it only depends on resources, all data is instantly available.

Reply:
Similarly, Watermark based triggers don't make sense either if all data is instantly available...

Reply:
Btw, the RDD runner does it the same way:
[https://github.com/apache/beam/blob/d578e3df7c963e57f251fb27739fbc1d3811e722/runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java#L144-L170](https://github.com/apache/beam/blob/d578e3df7c963e57f251fb27739fbc1d3811e722/runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java#L144-L170)

Reply:
i'll have a look at what the flink runner does as well.

Reply:
processing time trigger in batch mode is more a synomym of "not based on event time" rather than "based on wall clock".  In that case the trigger is based on elements: it is firing every 2 elements seen. So I think that makes sense in batch mode. 
Regarding watermak triggers in batch mode, they should better understood as triggers based on even timestamps as there is no flow of data in batch mode: please consider that event timestamps can be but on elements even in batch mode (cf replay of streaming recording) and in the case trigering on the event timestamp makes sense.  
But I agree these are corner cases not frequent at all among users.
Both nexmark queries run fine but beware nexmark does not check the correctness of data. So if there is no VR test on these corner cases, then it is not sure this behavior is correct.

Anyway fair enough as the RDD runner does the same and AFAIK no user complained.

Reply:
> processing time trigger in batch mode is more a synomym of "not based on event time" rather than "based on wall clock". In that case the trigger is based on elements: it is firing every 2 elements seen. So I think that makes sense in batch mode.

well, too some extend ... nevertheless, this is very artificial and hardly useful for the batch case. in your example it would be 2 random elements without any meaningful relationship (in the streaming use case that relationship is processing time).

Reply:
> Regarding watermak triggers in batch mode, they should better understood as triggers based on even timestamps as there is no flow of data in batch mode: please consider that event timestamps can be but on elements even in batch mode (cf replay of streaming recording) and in the case trigering on the event timestamp makes sense.

Event time is exactly the problem here ... assigning event time to elements is absolutely fine, but triggering based on event time is not possible. Data is not processed in a way that maintains event time (not even closely) at all unless there's some explicit ordering to ensure that. Event allowed lateness doesn't make sense in a batch context. Timestamp combiners, e.g. used with combiners on event timestamps, are absolutely valid on the other hand.

However, if supporting `RequiresTimeSortedInput` in Spark the situation probably looks different. In that case it probably makes sense to avoid the optimizations if a trigger is given and just use the DoFnRunner

Reply:
I'll give this a 2nd thought 👍

Reply:
I've had a closer look at `Distinct.values()` (and also added a test case for it) as it's implemented in terms of `Combine.PerKey` and explicitly checks the PaneInfo. That's working fine 👍